### ISSUE TITLE
local-exec provisioner: suppress sensitive output

### ISSUE THREAD

**Issue by tmaier** [hashicorp/terraform#17568]
I use the local-exec provisioner with a sensitive variable, e.g. "client_key"
```ruby
  provisioner "local-exec" {
    command = "echo '${var.client_key}' > ${path.module}/client_key"
  }
```

The terraform apply command would output the command and the result of the command to `stdin`.

I propose to add a `sensitive` argument:

```ruby
  provisioner "local-exec" {
    command = "echo '${var.client_key}' > ${path.module}/client_key"
    sensitive = true
  }
```

**Comment by danrabinowitz** [comment]
For remote-exec too, please.

**Comment by mwz** [comment]
+1

**Comment by apparentlymart** [comment]
In the mean time, perhaps passing the sensitive key via [`environment`](https://www.terraform.io/docs/provisioners/local-exec.html#environment) is a reasonable workaround? Should also sidestep any issues with shell escaping of any metacharacters that might be present in the value.

**Comment by pradeepbhadani** [comment]
@apparentlymart  `environment` can help to some extent but `local-exec` also put the bash output to logs which will contain the sensitive info

**Comment by liuml07** [comment]
+1

**Comment by ulm0** [comment]
Yes, please, this would be really appreciated. I'm trying as well to generate an ssh keypair within terraform, and populate a file with local_file resource, i just need the public key to be print to stdout, yet bot of them get printed.

```hcl
resource "tls_private_key" "ssh" {
  algorithm = "RSA"
  rsa_bits  = 4096
}

resource "local_file" "private_key" {
  count    = "${var.public_ssh_key == "" ? 1 : 0}"
  content  = "${tls_private_key.ssh.private_key_pem}"
  filename = "./keys/id_rsa"
}

resource "local_file" "public_key" {
  count    = "${var.public_ssh_key == "" ? 1 : 0}"
  content  = "${tls_private_key.ssh.public_key_openssh}"
  filename = "./keys/id_rsa.pub"
}

output "public_ssh_key" {
  # Only output a generated ssh public key
  value = "${var.public_ssh_key != "" ? "" : tls_private_key.ssh.public_key_openssh}"
}
```

Having a `sensitive = true` field would help

**Comment by subesokun** [comment]
Any progress on this issue?

**Comment by mayeco** [comment]
any progress??

**Comment by obikay200** [comment]
+1

**Comment by mayeco** [comment]
![image](https://user-images.githubusercontent.com/369490/57386612-b5207080-7182-11e9-9494-f47c18ee191a.png)

**Comment by apparentlymart** [comment]
Adding +1 comments doesn't achieve anything here because we can't report on them; instead, it just creates notification noise for everyone else who is following the issue. Instead, please add :+1: reactions to the original comment (not to _this_ comment), which we can and do [report on](https://github.com/hashicorp/terraform/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) as one input for prioritization.

**Comment by avthart** [comment]
Having a `sensitive = true` field would help

**Comment by darren-johnson** [comment]
Another vote here!

**Comment by moreirodamian** [comment]
+1

**Comment by omarSterckx** [comment]
+1

**Comment by JulianGogK24** [comment]
+1

**Comment by chad-barensfeld-exa** [comment]
+1

**Comment by iscooter** [comment]
+1

**Comment by mclavel** [comment]
```terraform
variable "super_secret" {
  type        = string
  description = "Use env: TF_VAR_super_secret"
}

resource "null_resource" "super_secret" {
  triggers = {
    hash_super_secret = sha256(var.super_secret)
  }

  provisioner "local-exec" {
    command = "echo $SUPER_SECRET >> /tmp/super_secret.txt"
    environment = {
      SUPER_SECRET = var.super_secret
    }
  }

  provisioner "local-exec" {
    when    = destroy
    command = "rm -f /tmp/super_secret.txt"
  }
}
```
@apparentlymart idea 😄

**Comment by alisdair** [comment]
This issue should be addressed as part of the work on sensitive variables and values in Terraform 0.14. When a provider configuration includes a sensitive value (from a sensitive variable, output, or resource attribute), [Terraform will suppress logging from the provisioner](https://github.com/hashicorp/terraform/pull/26611).

See this screenshot from the #26611 PR, based on the example from @tmaier's original report:

![Screenshot of suppressed provisioner output due to `sensitive` variable](https://user-images.githubusercontent.com/68917/96302627-1f1f8500-0fc7-11eb-956a-f684854a731a.png)

Because the `client_key` variable is marked sensitive, the derived `command` is also sensitive, and therefore the normal `local-exec` logging is suppressed.

If you're interested in this issue, please give this feature a try when you can. It's available in [the 0.14.0-beta2 release, out now](https://releases.hashicorp.com/terraform/0.14.0-beta2/). We're interested in hearing your feedback!
